PROTOCOL_V6_DRAFT.md

```diff
@@ -0,0 +1,201 @@
+# EXPERIMENTAL Twirp v6 Wire Protocol
+
+This document is a draft of a new version ("v6") of the Twirp wire
+protocol over HTTP. This draft is EXPERIMENTAL and subject to change.
+v6 of Twirp is not released or considered stable; instead, this
+document lists planned changes which may be available in prerelease
+distributions.
+
+## Changes from v5
+
+### URL scheme
+
+In [v5](./PROTOCOL.md), URLs followed this format:
+
+**URL ::= Base-URL "/twirp/" [ Package "." ] Service "/" Method**
+
+In v6, the "/twirp/" prefix is removed:
+
+**URL ::= Base-URL "/" [ Package "." ] Service "/" Method**
+
+The "/twirp/" prefix is removed for three reasons:
+
+ - Trademark concerns: some very large organizations don't want to
+   take any legal risks and are concerned that "twirp" could become
+   trademarked.
+ - Feels like advertising: To some users, putting "twirp" in all your
+   routes feels like it's just supposed to pump Twirp's brand, and
+   provides no value back to users.
+ - Homophonous with "twerp": In some Very Serious settings (like
+   government websites), it's not okay that "twirp" sounds like
+   "twerp", which means something like "insignificant pest."
+
+## Overview
+
+The Twirp wire protocol is a simple RPC protocol based on HTTP and
+Protocol Buffers (proto). The protocol uses HTTP URLs to specify the
+RPC endpoints, and sends/receives proto messages as HTTP
+request/response bodies.
+
+To use Twirp, developers first define their APIs using proto files,
+then use Twirp tools to generate the client and the server libraries.
+The generated libraries implement the Twirp wire protocol, using the
+standard HTTP library provided by the programming language runtime or
+the operating system. Once the client and the server are implemented,
+the client can communicate with the server by making RPC calls.
+
+The Twirp wire protocol supports both binary and JSON encodings of
+proto messages, and works with any HTTP client and any HTTP version.
+
+### URLs
+
+In [ABNF syntax](https://tools.ietf.org/html/rfc5234), Twirp's URLs
+have the following format:
+
+**URL ::= Base-URL "/" [ Package "." ] Service "/" Method**
```

Comment:
This URL schema is identical to gRPC, which seems confusing. At least it would conflict with existing logging, monitoring, routing, load balancing systems. While I am not saying gRPC should own this URL schema, it is better not to be identical between Twirp and gRPC.

Reply:
I thought the schema was supposed to be `URL ::= Base-URL "/" [ Package "/" ] Service "/" Method`